subiquity/tests/api/test_api.py

```diff
@@ -0,0 +1,521 @@
+#!/usr/bin/env python3
+
+import aiohttp
+from aiohttp.client_exceptions import ClientResponseError
+import async_timeout
+import asyncio
+from functools import wraps
+import json
+import logging
+import os
+import sys
+import unittest
+
+from subiquitycore.utils import astart_command
+
+
+logging.basicConfig(level=logging.DEBUG)
+socket_path = '.subiquity/socket'
+
+
+ver = sys.version_info
+if ver.major < 3 or ver.minor < 8:
+    raise Exception('skip asyncio testing')
+
+
+def find(items, key, value):
+    for item in items:
+        if item[key] == value:
+            yield item
+
+
+def first(items, key, value):
+    return next(find(items, key, value))
+
+
+def timeout(_timeout):
+    def wrapper(coro):
+        @wraps(coro)
+        async def run(*args, **kwargs):
+            with async_timeout.timeout(_timeout):
+                return await coro(*args, **kwargs)
+        return run
+    return wrapper
+
+
+def json_print(json_data):
+    print(json.dumps(json_data, indent=4))
+
+
+def loads(data):
+    return json.loads(data) if data else None
+
+
+def dumps(data):
+    # if the data we're dumping is literally False, we want that to be 'false'
+    if data or type(data) is bool:
+        return json.dumps(data, separators=(',', ':'))
+    elif data is not None:
+        return '""'
+    else:
+        return data
+
+
+class ClientException(Exception):
+    pass
+
+
+class TestAPI(unittest.IsolatedAsyncioTestCase):
+    machine_config = 'examples/simple.json'
+    need_spawn_server = True
+
+    async def get(self, query, **kwargs):
+        return await self.request('GET', query, **kwargs)
+
+    async def post(self, query, data=None, **kwargs):
+        return await self.request('POST', query, data, **kwargs)
+
+    async def request(self, method, query, data=None, **kwargs):
+        params = {}
+        for key in kwargs:
+            params[key] = dumps(kwargs[key])
+        data = dumps(data)
+        info = f'{method} {query}'
+        if params:
+            for i, key in enumerate(params):
+                joiner = '?' if i == 0 else '&'
+                info += f'{joiner}{key}={params[key]}'
+        print(info)
```

Comment:
I guess this is only for debugging but surely you don't have to print this by hand? Would printing resp.url below be good enough?

Reply:
Agree